### DISCUSSION TITLE
Whatsapp Video Message Request

### DISCUSSION THREAD

**Discussion by invictoring** [gabrieldwight/Whatsapp-Business-Cloud-Api-Net#20] [General]
I am able to send text message successfully but video message is not sending also guide me what is video url and do we need to first upload video to url then use that url. I tried youtube video url but that does not seem to be working. Kindly help me thanks.

**Comment by gabrieldwight** [comment]
Hi @invictoring if you are sending a youtube video URL you will need to use the text message with previewurl property set to true. To send a video message for the user to download, you must upload the video using **UploadMedia Function** which returns the **MediaId**. You can use **MediaId** to get the **media URL** from the **GetMediaUrl function**, take note the media URL returned from the cloud API is valid for **5 minutes** and the size limit is **16MB**. If the cloud API returns the **media URL** you can send the video for the user to download. The same applies to any media message you want to send the user. Let me know if you are able to send a video message.

**Reply by invictoring** [reply]
Thanks you so much for helping me out. I am now able to send youtube video successfully. Can you please share code for sending UploadMedia function and send video through that thanks.

**Reply by gabrieldwight** [reply]
For example to upload media using IFormFile
```C#
[HttpPost]
[ValidateAntiForgeryToken]
public async Task<IActionResult> UploadMedia(IFormFile mediaFile)
{
            try
            {
                var fileName = Path.GetFileName(mediaFile.FileName).Trim('"');

                var rootPath = Path.Combine(_environment.WebRootPath, "Application_Files\\MediaUploads\\");

                if (!Directory.Exists(rootPath))
                {
                    Directory.CreateDirectory(rootPath);
                }

                // Get the path of filename
                var filePath = Path.Combine(_environment.WebRootPath, "Application_Files\\MediaUploads\\", fileName);

                // Upload file to the browser
                using (var stream = new FileStream(filePath, FileMode.Create))
                {
                    await mediaFile.CopyToAsync(stream);
                }

                UploadMediaRequest uploadMediaRequest = new UploadMediaRequest();
                uploadMediaRequest.File = filePath;
                uploadMediaRequest.Type = mediaFile.ContentType;

                var uploadMediaResult = await _whatsAppBusinessClient.UploadMediaAsync(uploadMediaRequest);
                ViewBag.MediaId = uploadMediaResult.MediaId;
                return View().WithSuccess("Success", "Successfully upload media.");
            }
            catch (WhatsappBusinessCloudAPIException ex)
            {
                _logger.LogError(ex, ex.Message);
                return View().WithDanger("Error", ex.Message);
            }
}
```

To send video message with media Id
```C#
if (!string.IsNullOrWhiteSpace(sendMediaMessage.MediaId))
                        {
                            VideoMessageByIdRequest videoMessage = new VideoMessageByIdRequest();
                            videoMessage.To = sendMediaMessage.RecipientPhoneNumber;
                            videoMessage.Video = new MediaVideo();
                            videoMessage.Video.Id = sendMediaMessage.MediaId;
                            videoMessage.Video.Caption = sendMediaMessage.Message;

                            WhatsAppResponse results = await _whatsAppBusinessClient.SendVideoAttachmentMessageByIdAsync(videoMessage);
                        }
```

To get Media Url from Media Id
```C#
MediaUrlResponse mediaUrlResponse = await GetMediaUrlAsync(mediaId);
string mediaUrl = mediaUrlResponse.Url;
```

To send video message with Url
```C#
 if (!string.IsNullOrWhiteSpace(mediaUrl))
                        {
                            VideoMessageByUrlRequest videoMessage = new VideoMessageByUrlRequest();
                            videoMessage.To = sendMediaMessage.RecipientPhoneNumber;
                            videoMessage.Video = new MediaVideoUrl();
                            videoMessage.Video.Link = mediaUrl
                            videoMessage.Video.Caption = sendMediaMessage.Message;

                            WhatsAppResponse results = await _whatsAppBusinessClient.SendVideoAttachmentMessageByUrlAsync(videoMessage);
                        }
```

**Comment by invictoring** [comment]
Thank you so much for your support. i have another question can we use whatsapp business api for business what steps do we need to take kindly guide thanks.

Do we need Business Solution Provider or we can do this ourselves by using this api.

**Reply by gabrieldwight** [reply]
The WhatsApp Business API for business if you want to use additional services like a chatbot, customer care support, etc. For that, you will need a Business Solution Provider to manage WhatsApp services for you and integrate with their API. [WhatsApp BSP](https://www.facebook.com/business/partner-directory/search?solution_type=messaging&platforms=whatsapp). For cloud API it can support some of the BSP features like a chatbot but you will need to make your implementation design on how to handle receiving messages and replying back to the message.

**Reply by invictoring** [reply]
I just want to send broadcast messages to clients as business can we do this without BSP ??

**Reply by gabrieldwight** [reply]
You can achieve this without BSP though will need to create a list of message objects with a customized message that you want to send for each of them. In the loop, you can achieve the broadcast messages for multiple phone numbers. The BSP already supports broadcast messages on its platform.

**Reply by invictoring** [reply]
Ok thanks alot for your help.

**Comment by theCJMan** [comment]
Hi there
I first played with a Python Web automation, but now am moving over to API..
Thank you for your project
I have:
1. Functional FB Test account 
2. Sent text messages, Template(text), Template(Images) <- the Image I placed on a web site I have
3. Several approved templates including a VideoTemplate

I saw that you have this (similar) Vid Template AND you use a URL.
The Video (less than 7Mb) URL again is on a site I have -> Note, i did not use the whatsapp upload method because your example seems to suggest I can point to any URL, 
I get a 200 response, but the video template message never reaches me.

Please assist..

` [HttpPost]
 [Route("[action]")]
 public async Task<ActionResult<WhatsAppResponse>> SendTemplate_video_ParameterAsync(SendTemplate_media_ParameterPayload payload)
 {
     VideoTemplateMessageRequest videoTemplateMessage = new();
     videoTemplateMessage.To = payload.PhoneNumber;
     videoTemplateMessage.Template = new();
     videoTemplateMessage.Template.Name = payload.TemplateName;
     videoTemplateMessage.Template.Language = new();
     videoTemplateMessage.Template.Language.Code = LanguageCode.English_US;
     videoTemplateMessage.Template.Components = new List<VideoMessageComponent>()
         {
             new VideoMessageComponent()
             {
                 Type = "header",
                 Parameters = new List<VideoMessageParameter>()
                 {
                     new VideoMessageParameter()
                     {
                         Type = "video",
                         Video = new WhatsappBusiness.CloudApi.Messages.Requests.Video()
                         {
                             //Id = sendTemplateMessageViewModel.MediaId,
                             Link = payload.mediaURL // Link point where your document can be downloaded or retrieved by WhatsApp
                         }
                     }
                 },
             },
             new VideoMessageComponent()
             {
                 Type = "body",
                 Parameters = new List<VideoMessageParameter>()
                 {
                     new VideoMessageParameter()
                     {
                         Type = "text",
                         Text = "Video Information"
                     },
                     new VideoMessageParameter()
                     {
                         Type = "text",
                         Text = "Video Information"
                     },
                 }
             }
         };
                   
     var results = await _whatsAppBusinessClient.SendVideoAttachmentTemplateMessageAsync(videoTemplateMessage);
     return results;
 }`

**Reply by gabrieldwight** [reply]
Hi, if you are testing with test phone number from Facebook sometimes it makes one to send the hello world template message so that you can receive messages from the test phone number. 

You can also check if your video message template type was approved sometimes they restrict utility template message type.

**Reply by theCJMan** [reply]
Yes I still have a temp number to send from and I am still in Dev/ Test mode on FB
Yes the template was approved and it is a Marketing Template

the URL, can it point to my website file, or must I first upload the Video file to WhatsApp, then get the URL from WhatsApp?

**Reply by gabrieldwight** [reply]
The url can point to your website as long as the url file downloadable then it will work or you can use a cloud storage service if available

**Reply by theCJMan** [reply]
I can view the video when browsing to the url
https://corneliusmostert.com/PubDocs/VID_dafp.mp4

**Reply by theCJMan** [reply]
I tried an easier piece of code, but i also get a 200 response, but I do not receive any message.

 [HttpPost]
 [Route("[action]")]
 public async Task<ActionResult<WhatsAppResponse>> SendVideoAsync(SendVideoPayload sendVideoPayload)
 {
     VideoMessageByUrlRequest videoMessage = new VideoMessageByUrlRequest();
     videoMessage.To = sendVideoPayload.PhoneNumber;
     videoMessage.Video = new MediaVideoUrl();
     videoMessage.Video.Link = sendVideoPayload.mediaURL;

     var results = await _whatsAppBusinessClient.SendVideoAttachmentMessageByUrlAsync(videoMessage);        
     return results;

 }

**Reply by theCJMan** [reply]
testing the Image send with URL works, and Image Template with URL is also working.
The Image and Video has the same security rights on my server 0644

 [HttpPost]
 [Route("[action]")]
 public async Task<ActionResult<WhatsAppResponse>> SendVideoAsync(SendMediaURLPayload sendVideoPayload)
 {
     VideoMessageByUrlRequest videoMessage = new VideoMessageByUrlRequest();
     videoMessage.To = sendVideoPayload.PhoneNumber;
     videoMessage.Video = new MediaVideoUrl();
     videoMessage.Video.Link = sendVideoPayload.mediaURL;
     videoMessage.Video.Caption = sendVideoPayload.Message;

     var results = await _whatsAppBusinessClient.SendVideoAttachmentMessageByUrlAsync(videoMessage);        
     return results;

 }

**Reply by gabrieldwight** [reply]
@theCJMan I have seen the issue, so you will need to compress your video file size to be below **16 MB** for either getting media id from uploading the video or media url for your videos. The video permission from your site is fine. You can confirm if your endpoint is returning video/mp4 as the content type.

**Reply by theCJMan** [reply]
the file is less than 7Mb, and the file uploaded is MP4. So that should not be an issue

**Reply by theCJMan** [reply]
Lets recap:
1. I have a FB Dev account with a Temp Number (Test/Dev Mode), Permanent System ID
2. The vid is MP4 and only 5.34Mb with URL https://corneliusmostert.com/PubDocs/dafp.mp4
3. I can send Text to my Cell, I can send Pictures to my cell

**NOTE:**
I tried 2 other MP4 (NOT taken with my Cell) and they do go through!!!
I can now send Vids and Vid Templates....

**Thank you**

**Reply by theCJMan** [reply]
Quick question..
Initially when I automated the WhatsApp Web, I had this limitation on files, though, lately I was able to send larger files.
I see with the current API send on Vid, I get a 200 responce, but if the vid is larger than 16Mb then it does not arrive in my Phone..
Do you know if this Web vs API is because there is a new API or is it something else?

**Reply by gabrieldwight** [reply]
@theCJMan I have seen the issue with your video hosted on your site. I inspected your video file from the site and it uses codec H.265 which cloud API will reject in sending the video file. I converted to H.264 and it was able to send the video. The cloud API will support video that contains this metadata; H.264 video codec and AAC audio codec. 

### Your original video file codec information
![image](https://github.com/gabrieldwight/Whatsapp-Business-Cloud-Api-Net/assets/28096255/b263e26e-b8a2-4be2-b1cc-4396f9a60b12)

### Converted video file codec information
![image](https://github.com/gabrieldwight/Whatsapp-Business-Cloud-Api-Net/assets/28096255/65778bf5-1467-42e6-8bff-197292af0b9d)

**Reply by theCJMan** [reply]
Thank you, so you are saying the HEVC is the main issue, 
Also are you saying the Business APIs can send H264, or are you saying the Web APIs can, but not the Business APIs?

This is interesting information, thank you

**Reply by gabrieldwight** [reply]
Yes, it won't accept HEVC or H.265 video. That's the limitation from cloud API. [Supported Media Types](https://developers.facebook.com/docs/whatsapp/cloud-api/reference/media/#supported-media-types)

**Reply by theCJMan** [reply]
My humble apologies, I am asking 2 different questions.
1. It sounded form your reply above that the Web API can send H.265 but not HEVC, because after you converted it (As per screenshot above to H.265 only), you were able to send it
2. But it also sounds like the Business APIs does not allow H.265 AND also it does not allow HEVC.

Or am I missing something?

**Reply by gabrieldwight** [reply]
The cloud api will just send the video if it has mp4 file format but the API performs internally to check the video codec of the video file on their end before sending the message to the user. I converted the file to have H.264 video codec and I was able to send it and receive the video message.

The business API does not allow videos that are encoded in either HEVC or H.265 codec.


You will be required to send video message with H.264 codec using the business API.

**Comment by theCJMan** [comment]
> Quick question.. Initially when I automated the WhatsApp Web, I had this limitation on files, though, lately I was able to send larger files. I see with the current API send on Vid, I get a 200 responce, but if the vid is larger than 16Mb then it does not arrive in my Phone.. Do you know if this Web vs API is because there is a new API or is it something else?

Do you have an answer on this limitation Vs, why I can send larger videos using the WhatsApp Web, vs the Business APIs?

**Reply by gabrieldwight** [reply]
The file size limitation is on the business API implementation by WhatsApp since WhatsApp Web uses a different API than the main WhatsApp uses but still does compression before sending the video file.

**Comment by theCJMan** [comment]
O boy!!

I red your message 3 times, and still did not get it, until my daughter showed me the 4 vs 5 (last digit) of H.264 vs H265

Thanx...

⁣Get BlueMail for Android ​
